code/modules/integrated_electronics/core/assemblies.dm

```diff
@@ -151,37 +232,69 @@
 			return TRUE
 
 		if("remove_circuit")
-			var/obj/item/integrated_circuit/C = locate(params["ref"]) in contents
+			var/obj/item/integrated_circuit/C = locate(params["ref"]) in assembly_components
 			if(!istype(C))
 				return
 			C.remove(usr)
 			return TRUE
 
 		if("bottom_circuit")
-			var/obj/item/integrated_circuit/C = locate(params["ref"]) in contents
+			var/obj/item/integrated_circuit/C = locate(params["ref"]) in assembly_components
 			if(!istype(C))
 				return
 			// Puts it at the bottom of our contents
 			// Note, this intentionally does *not* use forceMove, because forceMove will stop if it detects the same loc
 			C.loc = null
 			C.loc = src
+
+		if("input_selection")
+			var/obj/item/integrated_circuit/input/C = locate(params["ref"]) in assembly_components
+			if(C.ask_for_input(usr,held_item,usr.a_intent))
+				return TRUE
+			else
+				var/obj/item/integrated_circuit/D = C
+				D.attackby_react(usr,held_item,usr.a_intent)
+				return TRUE
+
 	return FALSE
+
+// Input window data.
+
 // End TGUI
+/* TBI	diag hud
+/obj/item/electronic_assembly/pickup(mob/living/user)
+	. = ..()
+	//update diagnostic hud when picked up, true is used to force the hud to be hidden
+	diag_hud_set_circuithealth(TRUE)
+	diag_hud_set_circuitcell(TRUE)
+	diag_hud_set_circuitstat(TRUE)
+	diag_hud_set_circuittracking(TRUE)
+
+/obj/item/electronic_assembly/dropped(mob/user)
+	. = ..()
+	//update diagnostic hud when dropped
+	diag_hud_set_circuithealth()
+	diag_hud_set_circuitcell()
+	diag_hud_set_circuitstat()
+	diag_hud_set_circuittracking()
 
+*/
 /obj/item/electronic_assembly/verb/rename()
 	set name = "Rename Circuit"
 	set category = "Object"
 	set desc = "Rename your circuit, useful to stay organized."
 
 	var/mob/M = usr
+	var/input = sanitizeSafe(input(usr, "What do you want to name this?", "Rename", src.name) as null|text, MAX_NAME_LEN)
```

Comment:
```suggestion
	var/input = tgui_input_text(usr, "What do you want to name this?", "Rename", src.name)
```
`tgui_input_text` handles sanitizing and deals with max len for ya :P

Reply:
Oh my god this will change my life, thank you!! fwiw that code was already there, I'm not sure if I even touched it. :P